src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -31,109 +30,134 @@
-                }
-                if (matchingOwners.size() > 1) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not unique");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                pet.setOwner(matchingOwners.iterator().next());
+            switch (entry.operation.toLowerCase()) {
+                case "add":
+                    clinicService.savePet(pet);
+                case "delete":
+                    // originally provided logic can be wrapped in this function
+                    clinicService.deletePet(pet);
             }
+            pets.add(pet);
+        }
+    }
 
-            if (csv.charAt(i) == ';') {
-                i++;
+    private HttpHeaders constructHttpHeaders(String headerValue){
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("errors", headerValue);
+        return headers;
+    }
 
-                field = "";
-                while (i < csv.length() && csv.charAt(i) != '\n') {
-                    field += csv.charAt(i++);
-                }
+    private Serializable toPetName(String name) {
+        if (name != "null") {
+            return name;
+        } else {
+            return constructHttpHeaders("Pet name is empty");
+        }
+    }
 
-                if (field.toLowerCase().equals("add")) {
-                    clinicService.savePet(pet);
-                } else {
-                    for (Pet q : pet.getOwner().getPets()) {
-                        if (q.getName().equals(pet.getName())) {
-                            if (q.getType().getId().equals(pet.getType().getId())) {
-                                if (pet.getBirthDate().equals(q.getBirthDate())) {
-                                    clinicService.deletePet(q);
-                                }
-                            }
-                        }
-                    }
-                }
-
-            } else {
-                clinicService.savePet(pet);
-            }
-            i++;
+    private Serializable toDate(String date) {
+        try {
+            return (new SimpleDateFormat("yyyy-MM-dd")).parse(date);
+        } catch (ParseException e) {
+            return constructHttpHeaders("date " + date + " not valid");
+        }
+    }
 
-            pets.add(pet);
+    private Object toPetType(String petTypeCSV) {
+        ArrayList<PetType> petTypes = (ArrayList<PetType>) clinicService.findPetTypes();
+        for (PetType petType : petTypes) {
+            if (petType.getName().equalsIgnoreCase(petTypeCSV)) {
+                return petType;
+            }
+        }
+        return constructHttpHeaders("Pet type does not exist");
+    }
 
-        } while (i < csv.length() && pet != null);
+    private Object toOwner(String name) {
+        List<Owner> matchingOwners = clinicService.findAllOwners()
```

Comment:
this is quite dangerous for several reasons:
1. let the database do the filtering for you, they are highly optimized
2. this method is called for all  ```n``` entries of the csv list, and then you loop over all  ```n``` owners, so this is  ```O(n^2)```, a quadratic progress and can result in very bad performance (see https://en.wikipedia.org/wiki/Big_O_notation)